onmt/translate/predictor.py

```diff
@@ -0,0 +1,337 @@
+
+    classifier = Classifier.from_opt(
+        model,
+        fields,
+        opt,
+        model_opt,
+        out_file=out_file,
+        logger=logger
+    )
+    return classifier
+
+
+def build_tagger(opt, logger=None, out_file=None):
+    """Return a tagger with result redirect to `out_file`."""
+
+    if out_file is None:
+        out_file = codecs.open(opt.output, 'w+', 'utf-8')
+
+    load_bert_model = onmt.model_builder.load_bert_model
+    fields, model, model_opt = load_bert_model(opt, opt.model)
+
+    tagger = Tagger.from_opt(
+        model,
+        fields,
+        opt,
+        model_opt,
+        out_file=out_file,
+        logger=logger
+    )
+    return tagger
+
+
+class Predictor(object):
+    """Predictor a batch of data with a saved model.
+
+    Args:
+        model (nn.Sequential): model to use
+        fields (dict[str, torchtext.data.Field]): A dict of field.
+        gpu (int): GPU device. Set to negative for no GPU.
+        data_type (str): Source data type.
+        verbose (bool): Print/log every predition.
+        report_time (bool): Print/log total time/frequency.
+        out_file (TextIO or codecs.StreamReaderWriter): Output file.
+        logger (logging.Logger or NoneType): Logger.
+    """
+
+    def __init__(
+            self,
+            model,
+            fields,
+            gpu=-1,
+            verbose=False,
+            out_file=None,
+            report_time=True,
+            logger=None,
+            seed=-1):
+        self.model = model
+        self.fields = fields
+
+        self._gpu = gpu
+        self._use_cuda = gpu > -1
+        self._dev = torch.device("cuda", self._gpu) \
+            if self._use_cuda else torch.device("cpu")
+
+        self.verbose = verbose
+        self.report_time = report_time
+        self.out_file = out_file
+        self.logger = logger
+
+        set_random_seed(seed, self._use_cuda)
+
+    @classmethod
+    def from_opt(
+            cls,
+            model,
+            fields,
+            opt,
+            model_opt,
+            out_file=None,
+            logger=None):
+        """Alternate constructor.
+
+        Args:
+            model (onmt.modules): See :func:`__init__()`.
+            fields (dict[str, torchtext.data.Field]): See
+                :func:`__init__()`.
+            opt (argparse.Namespace): Command line options
+            model_opt (argparse.Namespace): Command line options saved with
+                the model checkpoint.
+            out_file (TextIO or codecs.StreamReaderWriter): See
+                :func:`__init__()`.
+            logger (logging.Logger or NoneType): See :func:`__init__()`.
+        """
+
+        return cls(
+            model,
+            fields,
+            gpu=opt.gpu,
+            verbose=opt.verbose,
+            out_file=out_file,
+            logger=logger,
+            seed=opt.seed)
+
+    def _log(self, msg):
+        if self.logger:
```

Comment:
You are checking whether logger is `None` or not right? Let's do it explicitly `if self.logger is None:`